src/psyclone/tests/psyir/nodes/dynamic_omp_task_directive_test.py

```diff
@@ -0,0 +1,3915 @@
+    of a written array is a shared variable.'''
+    code = '''
+    subroutine my_subroutine()
+        type :: x
+          integer, dimension(320, 10) :: A
+        end type
+        type(x) :: AA
+        integer, dimension(321, 10) :: B
+        integer :: i
+        integer :: j
+        integer :: k
+        do i = 1, 320, 32
+            k = 4
+            k = -2
+            k = k + 3
+            do j = 1, 32
+                AA%A(i, k) = k
+                AA%A(i, j) = B(1+i, j) + k
+            end do
+        end do
+    end subroutine
+    '''
+    tree = fortran_reader.psyir_from_source(code)
+    ptrans = OMPParallelTrans()
+    strans = OMPSingleTrans()
+    tdir = DynamicOMPTaskDirective()
+    loops = tree.walk(Loop, stop_type=Loop)
+    loop = loops[0].children[3].children[3]
+    parent = loop.parent
+    loop.detach()
+    tdir.children[0].addchild(loop)
+    parent.addchild(tdir, index=0)
+    strans.apply(loops[0])
+    ptrans.apply(loops[0].parent.parent)
+
+    with pytest.raises(GenerationError) as excinfo:
+        tree.lower_to_language_level()
+    assert ("Shared variable access used as an array index inside an "
+            "OMPTaskDirective which is not supported. Variable name is "
+            "'Reference[name:'k']'" in str(excinfo.value))
+
+
+def test_omp_task_directive_read_index_shared_type(fortran_reader):
+    ''' Test the code generation correctly generates an error if an array index
+    of a read array is a shared variable.'''
+    code = '''
+    subroutine my_subroutine()
+        type :: x
+          integer, dimension(320, 10) :: A
+        end type
+        type(x) :: AA
+        integer, dimension(321, 10) :: B
+        integer :: i
+        integer :: j
+        integer :: k
+        do i = 1, 320, 32
+            k = 4
+            k = -2
+            k = k + 3
+            do j = 1, 32
+                AA%A(i, j) = AA%A(i, k)
+                AA%A(i, j) = B(1+i, j) + k
+            end do
+        end do
+    end subroutine
+    '''
+    tree = fortran_reader.psyir_from_source(code)
+    ptrans = OMPParallelTrans()
+    strans = OMPSingleTrans()
+    tdir = DynamicOMPTaskDirective()
+    loops = tree.walk(Loop, stop_type=Loop)
+    loop = loops[0].children[3].children[3]
+    parent = loop.parent
+    loop.detach()
+    tdir.children[0].addchild(loop)
+    parent.addchild(tdir, index=0)
+    strans.apply(loops[0])
+    ptrans.apply(loops[0].parent.parent)
+    with pytest.raises(GenerationError) as excinfo:
+        tree.lower_to_language_level()
+    assert ("Shared variable access used as an array index inside an "
+            "OMPTaskDirective which is not supported. Variable name is "
+            "'Reference[name:'k']'" in str(excinfo.value))
+
+
+def test_omp_task_directive_33(fortran_reader):
+    ''' Test the code generation fails when attempting to access an array
+    when using an array element as an index.'''
```

Comment:
Again, needs extra detail (and perhaps move this test so that it's next to similar previous ones).

Reply:
Added, and moved more tests around that make sense to be near each other.